feat(check_port): Implement dual IPv4/IPv6 port status checking and display

cantalupo555 opened the pull request:
- Detect public IPv4 and IPv6 addresses using dedicated APIs (ipify.org).
- Perform port status checks for both IPv4 and IPv6 protocols independently.
- Update the backend (action.php) to return a JSON response containing status for both protocols.
- Revise the frontend to display separate status icons in the ruTorrent status bar, with each IP address shown next to its respective icon.
- Update the tooltip on the status bar pane to show detailed port status for both IPv4 and IPv6.
- Improve visual layout with proper spacing and alignment for status indicators.

cantalupo555 committed:
feat(check_port): Implement dual IPv4/IPv6 port status checking and display

- Detect public IPv4 and IPv6 addresses using dedicated APIs (ipify.org).
- Perform port status checks for both IPv4 and IPv6 protocols independently.
- Update the backend (action.php) to return a JSON response containing status for both protocols.
- Revise the frontend to display separate status icons in the ruTorrent status bar, with each IP address shown next to its respective icon.
- Update the tooltip on the status bar pane to show detailed port status for both IPv4 and IPv6.
- Improve visual layout with proper spacing and alignment for status indicators.

stickz commented:
Thanks for rewriting check port. 

I would suggest creating two plugins. One for IPV4 and one for IPV6. This will give more control to toggle the features. We can't relay on the configuration file. Some common platforms (like hosted environments) do not allow users to access it.

Anther thing that would be really cool for IPV4 is a fallback option. We could automatically fallback to `portchecker` if `yougetsignal` goes offline. This would greatly improve the reliability of the feature for IPV4.

cantalupo555 commented:
> I would suggest creating two plugins. One for IPV4 and one for IPV6.

@stickz 

Yes, at first I seriously considered creating two separate plugins to handle this new implementation. After all, there were several possible approaches to explore. However, I was uncertain whether introducing two new plugins—and potentially retiring `check_port` or merging it with another plugin capable of IPv6 verification—would be the best solution.

> We can't rely on the configuration file. Some common platforms (like hosted environments) do not allow users to access it.

To keep things straightforward, I ultimately decided to consolidate everything into a single plugin. That said, I’d like to better understand your point about the configuration file’s reliability. Since I preserved the same configuration structure as the original plugin, I believe it could be enhanced to adapt to various environments—such as cases where a service, IPv4, or IPv6 is unavailable. Could you clarify in which specific situations the configuration file might not be trustworthy? This would help me address any potential limitations more effectively.

stickz commented:
@cantalupo555 These are very good points.

1. I would suggest using the current `check_port` plugin for IPV4. Then creating a new `check_port2` plugin for IPV6.

2. The limitation with the configuration file is that we cannot control the feature from the web interface.

stickz commented:
@cantalupo555 Any update here? I've set a milestone for ruTorrent `v5.3`. After this PR is completed, I plan to release it.

cantalupo555 commented:
> @cantalupo555 Any update here? I've set a milestone for ruTorrent `v5.3`. After this PR is completed, I plan to release it.

Hello @stickz
I intend to release it this week.
Believe it is possible to release it together with version 5.3

stickz commented:
> > @cantalupo555 Any update here? I've set a milestone for ruTorrent `v5.3`. After this PR is completed, I plan to release it.
> 
> Hello @stickz I intend to release it this week. Believe it is possible to release it together with version 5.3

Awesome thank you so much. I will release it together with `v5.3`.

cantalupo555 commented:
@stickz I apologize for the delayed response. Over the past few days, I’ve had some time to reflect on the implementation.

I’m concerned about code duplication and the complexity it may create for maintenance. We could work on improvements to make `check_port` smarter by implementing features such as:

- Automatic service detection;
- Fallbacks for IP retrieval;
- Fallbacks for checking services;
- Automatic checks or result caching, which would also address the limitation of configuration via the web interface.

Additionally, it would be valuable to make it more intelligent by enabling the plugin to automatically disable checking if only one of the two protocols is available.

I’m increasingly concerned about code duplication and its implications.

stickz commented:
> @stickz I apologize for the delayed response. Over the past few days, I’ve had some time to reflect on the implementation.
> 
> I’m increasingly concerned about code duplication and its implications.

Okay, we can find a way to keep this in one plugin. It will be more complicated though.

>  We could work on improvements to make `check_port` smarter by implementing features such as:
> 
> * Automatic service detection;
> * Fallbacks for IP retrieval;
> * Fallbacks for checking services;
> * Automatic checks or result caching, which would also address the limitation of configuration via the web interface.
> 

Let's merge this pull request. We can make a new pull request for additional features.

stickz commented on plugins/check_port/action.php:
```diff
@@ -1,142 +1,194 @@
+
+// Send the final JSON response to the client.
+header('Content-Type: application/json');
+echo json_encode($response);
```
Please use `CachedEcho::send(JSON::safeEncode($response),"application/json");` These two implementations are standard across all plugins using the two wrappers.

cantalupo555 replied:
I'm looking into this.

stickz replied:
@cantalupo555 I'm waiting for this line to be changed before I can test and possibly merge.

cantalupo555 replied:
@stickz Two new commits
https://github.com/Novik/ruTorrent/pull/2957/commits/98900bb6626c734f98f9acc65151cbb25e86a0e1
https://github.com/Novik/ruTorrent/pull/2957/commits/cdc44aac4441c5759f9cad6754cbb287727bdf15

stickz commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+
+// Creates and shows the context menu (right-click menu) for the status pane.
+plugin.createPortMenu = function(e) {
+ 	if (e.which === 3) { // Right mouse button.
```
Spacing issue.

cantalupo555 replied:
I applied standardization to the spacing (indentation) using TAB, consistent with the rest of the project. I usually use spaces, but the next commit will align with what seems to be the ruTorrent convention.

cantalupo555 replied:
@stickz https://github.com/Novik/ruTorrent/pull/2957/commits/f9d8dc212025bfeccb677f7c239cf33140849daa
I need you to check again

stickz replied:
Spacing looks good thanks.

stickz commented on plugins/check_port/action.php:
```diff
@@ -1,142 +1,194 @@
+    } else {
+        error_log("check_port plugin: Failed to fetch from {$url}. Status: {$snoopy->status}, Error: {$snoopy->error}");
+    }
+ 	return null; // Return null on failure.
```
spacing issue

stickz commented on plugins/check_port/action.php:
```diff
@@ -1,142 +1,194 @@
+ * @param int $timeout Request timeout.
+ * @return string|null The public IP address or null on failure.
+ */
+function get_public_ip($version, $timeout) {
+    if (!Utility::getExternal('curl')) {
+        error_log("check_port plugin: 'curl' executable not found.");
+        return null;
+    }
+ 	// Initialize the Snoopy client.
+    $snoopy = new Snoopy();
+    $snoopy->agent = "ruTorrent CheckPort Plugin/IP Check";
+ 	// Set a timeout for the request, with a minimum of 5 seconds.
+    $snoopy->read_timeout = max(5, (int)($timeout / 2));
+ 	$snoopy->proxy_host = ""; // Do not use a proxy for this external IP check.
+
+ 	// Select the correct ipify API URL based on the requested IP version.
+    $url = ($version == '6') ? "https://api64.ipify.org/" : "https://api4.ipify.org/";
+ 	@$snoopy->fetch($url); // Fetch the URL.
+
+ 	// Check if the request was successful and returned content.
+    if ($snoopy->status == 200 && !empty($snoopy->results)) {
+        $ip = trim($snoopy->results);
+ 		// Validate the returned IP address to ensure it's a valid IPv4 or IPv6.
```
Various spacing issues.

cantalupo555 replied:
This spacing problem must have been in my IDE 😱

stickz commented on plugins/check_port/action.php:
```diff
@@ -1,142 +1,194 @@
+ 		// Validate the returned IP address to ensure it's a valid IPv4 or IPv6.
+        $flag = ($version == '6') ? FILTER_FLAG_IPV6 : FILTER_FLAG_IPV4;
+        if (filter_var($ip, FILTER_VALIDATE_IP, $flag)) {
+ 			return $ip; // Return the valid IP.
```
spacing issue.

stickz commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+ 	// Set a tooltip to indicate that a check is in progress.
+    let title = theUILang.checkingPort || "Checking port status...";
+    if (isUpdate) {
+ 		title += "..."; // Append ellipsis for manual updates.
```
spacing issue.

stickz commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+    a.separator.text("").hide();
+    a.textIPv6.text("").hide();
+    
+ 	// Set a tooltip to indicate that a check is in progress.
```
various spacing issues.

stickz commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+    const port = data[proto + '_port'];
+ 	const isAvailable = address && address !== "-"; // Check if an IP address was returned.
+    
+ 	// Select the correct UI elements for the given protocol.
```
spacing issues.

stickz commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+        a.pane.on("mousedown", plugin.createPortMenu);
+    }
+    
+ 	// Trigger the initial port check.
```
Various spacing issues.

stickz commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+
+// This function is called when the plugin is removed/unloaded.
+plugin.onRemove = function() {
+ 	// Remove the pane from the status bar to clean up the UI.
```
spacing issue.

stickz reviewed (changes requested):
Please resolve spacing issues to ensure proper code alignment.

Also, please use our standard wrappers for
1) Echoing content in a PHP file.
2) Encoding content into JSON format.

Afterwards, I will test and review the plugin rewrite.

cantalupo555 commented:
Thanks for the review, I'm working on it...

cantalupo555 commented:
> Let's merge this pull request. We can make a new pull request for additional features.

Certainly, the plugin is still not at the level I aim to achieve.

Thanks for the review, I'm working on it...

cantalupo555 commented:
~~I'm renaming the origin repository branch to better align with convention.~~

cantalupo555 committed:
refactor(check_port): Standardize indentation to use tabs

This commit updates the code formatting to use tabs for indentation, replacing spaces in several files.
The change aligns the code with the predominant style of the ruTorrent, ensuring greater consistency across the codebase.

cantalupo555 commented on plugins/check_port/init.js:
```diff
@@ -1,68 +1,170 @@
+	icon.removeClass("pstatus0 pstatus1 pstatus2").addClass("pstatus" + status);
+
+	let displayText = "";
+	let titleText = "";
```
`let titleText = "";`
This now seems redundant to me.

stickz replied:
Not important, you can leave it.

cantalupo555 committed:
refactor(check_port): Standardize JSON response handling

This change brings the `check_port` plugin in line with the established coding practices, improving code consistency.

cantalupo555 committed:
docs(check_port): Improve comment readability and consistency

Removes trailing periods from comments throughout the plugin's files. This change standardizes the comment style, making the code easier to read and maintain.

stickz commented on plugins/check_port/conf.php:
```diff
@@ -1,9 +1,15 @@
+									// "portchecker" - use https://portchecker.co/
 
-$useIpv4 = true;		// use ipv4 addresses when checking ports, unless website supports ipv6
+$useWebsiteIPv6 = "portchecker";	// Valid choices:
```
We need to set this to false for this pull request. I have IPV6 disabled. It's still going off.
This issue needs to be fixed in a separate pull request, for documentation purposes.

cantalupo555 replied:
So, I need to anticipate the new feature.

cantalupo555 replied:
@stickz It's ready, I just need you to merge this pull request so we can continue.

I already have two new commits.
https://github.com/cantalupo555/ruTorrent/commit/d1ea3f1310a104c29955d166e9e81177229e1736
https://github.com/cantalupo555/ruTorrent/commit/c928f490481aebf99abd7cfaa88069849c85de12

stickz reviewed (changes requested):
IPV6 icon does not hide when IPV6 checking is disabled. Error log observed.
`check_port plugin: https://api64.ipify.org/ returned invalid IP: xxx.xxx.xxx.xxx`

cantalupo555 commented:
> Additionally, it would be valuable to make it more intelligent by enabling the plugin to automatically disable checking if only one of the two protocols is available.

I'm working on this because of the appeal...
To be able to reach version `v5.3`.

cantalupo555 commented:
![Screenshot from 2025-07-09 19-25-49](https://github.com/user-attachments/assets/98d336f4-6d79-45c2-a9fb-cd701ba11470)

We will have 2 new explicit parameters.
`use_ipv4` and `use_ipv6`

> This issue needs to be fixed in a separate pull request, for documentation purposes.

Perfect, I'm preparing it on a new branch.
`feature/check_port/conditional-protocol-display`


---

@stickz Thanks for following the progress; it helps a lot.

cantalupo555 commented:
@stickz I saw the first PR was merged using 'squash'. Since my second PR was dependent on it, I'll need to manually update my branch to align the history. 🙄

stickz commented:
> @stickz I saw the first PR was merged using 'squash'. Since my second PR was dependent on it, I'll need to manually update my branch to align the history. 🙄
> 
You should be able to just rebase with a few clicks.

cantalupo555 commented:
> > @stickz I saw the first PR was merged using 'squash'. Since my second PR was dependent on it, I'll need to manually update my branch to align the history. 🙄
> 
> You should be able to just rebase with a few clicks.

The downside of "squashing" is that my profile stats will receive less data in the activity graph. 😰

usr20210909 commented:
> IPV6 icon does not hide when IPV6 checking is disabled. Error log observed.
> `check_port plugin: https://api64.ipify.org/ returned invalid IP: xxx.xxx.xxx.xxx`

This is still the case in v5.3.0 release.
